### PR TITLE
[IRGen/Distributed] Adjust protocol requirement and witness table to handle distributed thunks

### PR DESCRIPTION
**Pull request by xedin**
- Determine whether protocol requirement is async using `SILDeclRef`

   This is necessary because SILDeclRef of the distributed thunk witness is anchored
   on the requirement declaration which might not be async, but the witness itself is
   always `async throws`.

-  Mark protocol requirement descriptor for distributed thunk as async

   Fixes a bug where descriptor for protocol requirement associated with
   distributed thunk wasn't marked as `async` that results in ptrauth failures
   because discriminator would be misaligned between requirement and witness.

Resolves: rdar://96520492
Resolves: rdar://97228310

### CODE REVIEW

**Comment by xedin** [comment]
@swift-ci please test

**Comment by xedin** [comment]
@swift-ci please test

**Comment by xedin** [comment]
@swift-ci please test apple silicon

**Comment by xedin** [comment]
@swift-ci please test apple silicon

File: lib/IRGen/GenMeta.cpp [hunk 1]
```diff
@@ -913,6 +913,10 @@ namespace {
         if (entry.getFunction().isAutoDiffDerivativeFunction())
           declRef = declRef.asAutoDiffDerivativeFunction(
               entry.getFunction().getAutoDiffDerivativeFunctionIdentifier());
+        if (entry.getFunction().isDistributedThunk()) {
+          flags = flags.withIsAsync(true);
+          declRef = declRef.asDistributed();
+        }
```
Comment (ktoso) [inline lib/IRGen/GenMeta.cpp:919]: okey so this piece we did yesterday did matter... 👍

File: lib/IRGen/GenProto.cpp [hunk 2]
```diff
@@ -1388,8 +1389,9 @@ class AccessorConformanceInfo : public ConformanceInfo {
       witness = llvm::ConstantExpr::getBitCast(witness, IGM.Int8PtrTy);
 
       PointerAuthSchema schema =
-          afd->hasAsync() ? IGM.getOptions().PointerAuth.AsyncProtocolWitnesses
-                          : IGM.getOptions().PointerAuth.ProtocolWitnesses;
+          isAsyncRequirement
```
Comment (ktoso) [inline lib/IRGen/GenProto.cpp:1392]: aha that's what we were missing I see 👀

**Comment by ktoso** [comment]
Apple silicon failing to start, times out;

**Comment by ktoso** [comment]
@swift-ci please test apple silicon

**Comment by ktoso** [comment]
Also resolves rdar://97228310

**Comment by xedin** [comment]
https://github.com/apple/swift-package-manager/pull/5684
@swift-ci please test apple silicon

**Comment by ktoso** [comment]
The revert got merged so the above AS run failed:

```

/Users/ec2-user/jenkins/workspace/swift-pr-arm64-macos/branch-main/swiftpm failed (ret=128): ['git', 'fetch', 'origin', 'pull/5684/merge:ci_pr_5684', '--tags']
fatal: couldn't find remote ref pull/5684/merge

update-checkout failed, fix errors and try again
```

**Comment by ktoso** [comment]
@swift-ci please test apple silicon

**Comment by xedin** [comment]
@swift-ci please test apple silicon

**Pull request merged**